textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/keywordmatch/KeywordMatcher.java

```diff
@@ -124,10 +140,62 @@ public ITuple getNextTuple() throws DataFlowException {
                     }
 
                     boolean allTokenPresent = areAllTrue(tokensPresent);
+                    if(predicate.getOperatorType() == KeywordOperatorType.BASIC) {
+
+                        if (!allTokenPresent) {
+                            spanList.removeAll(spanForThisField);
+                        }
+                    }
+
+                    else if(predicate.getOperatorType() == KeywordOperatorType.PHRASE){
 
-                    if (!allTokenPresent) {
                         spanList.removeAll(spanForThisField);
+                        //relevantWordsInQuery has all the tokens (including duplicates but excluding stop words) in the query.
```

Comment:
Can we check `predicate.getOperatorType() == KeywordOperatorType.PHRASE` earlier so that we don't need to even compute `spanForThisField` for the PHRASE case?